framework/src/main/java/org/checkerframework/common/value/PropertyFileHandler.java

```diff
@@ -0,0 +1,331 @@
+        return superAnnoElementValue.equals(subAnnoElementValue);
+    }
+
+    /**
+     * Tests whether subAnno is subtype of superAnno in the property file type hierarchy.
+     *
+     * @param subAnno the sub qualifier
+     * @param superAnno the super qualifier
+     * @return true if subAnno is subtype of superAnno
+     */
+    protected boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+        if (AnnotationUtils.areSameByClass(subAnno, PropertyFileBottom.class)
+                || AnnotationUtils.areSameByClass(superAnno, PropertyFileUnknown.class)) {
+            return true;
+        } else if (AnnotationUtils.areSameByClass(subAnno, PropertyFileUnknown.class)
+                || AnnotationUtils.areSameByClass(superAnno, PropertyFileBottom.class)) {
+            return false;
+        } else if (AnnotationUtils.areSameByClass(subAnno, PropertyFile.class)
+                && AnnotationUtils.areSameByClass(superAnno, PropertyFile.class)) {
+            return comparePropFileElementValue(subAnno, superAnno);
+        } else {
+            throw new BugInCF("We should never reach here.");
+        }
+    }
+
+    /**
+     * Try to read the value of the key in the provided property file.
+     *
+     * @param propFile the property file to open
+     * @param key the key to find in the property file
+     * @param defaultValue the default value of that key
+     * @return the value of the key in the property file
+     */
+    protected String readValueFromPropertyFile(String propFile, String key, String defaultValue) {
+        String res = null;
+        try {
+            Properties prop = new Properties();
+            ClassLoader cl = this.getClass().getClassLoader();
+            if (cl == null) {
+                // the class loader is null if the system class loader was
+                // used
+                cl = ClassLoader.getSystemClassLoader();
+            }
+            InputStream in = cl.getResourceAsStream(propFile);
+            if (in == null) {
+                // if the classloader didn't manage to load the file, try
+                // whether a FileInputStream works. For absolute paths this
+                // might help.
+                try {
+                    in = new FileInputStream(propFile);
+                } catch (FileNotFoundException e) {
+                    // ignore
+                }
+            }
+            if (in == null) {
+                checker.message(Kind.WARNING, "Couldn't find the properties file: " + propFile);
+                return null;
+            }
+            prop.load(in);
+            if (defaultValue == null) {
+                res = prop.getProperty(key);
+            } else {
+                res = prop.getProperty(key, defaultValue);
+            }
+        } catch (Exception e) {
+            checker.message(
+                    Kind.WARNING, "Exception in PropertyFileHandler.readPropertyFromFile: " + e);
+            e.printStackTrace();
```

Comment:
Remove.

Reply:
Same reason here. We cannot use `reportWarning` because there is no AST node to be passed to `reportWarning`. Same issue in the propkey checker: https://github.com/typetools/checker-framework/blob/master/checker/src/main/java/org/checkerframework/checker/propkey/PropertyKeyAnnotatedTypeFactory.java#L189